From e09228761dbaa5ba3a0bf56d1bbb96c867f7207d Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 28 Apr 2022 10:52:37 +0300
Subject: [PATCH] [SW-82716] common: add timestamp apis

In order to take care of synchronization issues related to i2c OOB
communication, this patch add timestamp api which will be used
by i2c OOB comms. The timestamp used is PSOC timestamp which take
clock from div3 of PCIE/PSOC PLL @ 50MHz.

Register map and mask are added for greco and gaudi2 to avoid
compilation issues.

Change-Id: I5b8cd2888b9ddbb358265fae9591f684c666f8a6
---

diff --git a/arch/gaudi2/arc/regmap.h b/arch/gaudi2/arc/regmap.h
index 93143d2..2b44124 100644
--- a/arch/gaudi2/arc/regmap.h
+++ b/arch/gaudi2/arc/regmap.h
@@ -17,6 +17,12 @@
 #define mmQSPI_BAUDR			mmPSOC_QSPI_BAUDR
 #define mmQSPI_DR0			mmPSOC_QSPI_DR0
 
+#define mmTIMESTAMP_CNTCR		mmPSOC_TIMESTAMP_CNTCR
+#define mmTIMESTAMP_CNTSR		mmPSOC_TIMESTAMP_CNTSR
+#define mmTIMESTAMP_CNTCVL		mmPSOC_TIMESTAMP_CNTCVL
+#define mmTIMESTAMP_CNTCVU		mmPSOC_TIMESTAMP_CNTCVU
+#define mmTIMESTAMP_CNTFID0		mmPSOC_TIMESTAMP_CNTFID0
+
 #define mmEFUSE_READ_CTRL		mmPSOC_EFUSE_READ_CTRL
 #define mmEFUSE_READ_ADDRESS		mmPSOC_EFUSE_READ_ADDRESS
 #define mmEFUSE_DATA_0			mmPSOC_EFUSE_DATA_0
diff --git a/arch/gaudi2/arc/regmask.h b/arch/gaudi2/arc/regmask.h
index 4a1e217..070a33b 100644
--- a/arch/gaudi2/arc/regmask.h
+++ b/arch/gaudi2/arc/regmask.h
@@ -6,6 +6,9 @@
 #ifndef _GAUDI2_REGMASK_H_
 #define _GAUDI2_REGMASK_H_
 
+#define TIMESTAMP_CNTCR_CNTCR_HDBG_M		CPU_TIMESTAMP_CNTCR_CNTCR_HDBG_MASK
+#define TIMESTAMP_CNTCR_CNTCR_EN_M		CPU_TIMESTAMP_CNTCR_CNTCR_EN_MASK
+
 #define SPI_DMA_CTRL_ADDR_SHIFT			PSOC_GLOBAL_CONF_SPI_DMA_CTRL_ADDR_SHIFT
 #define SPI_DMA_CTRL_DST_SRAM_SHIFT		PSOC_GLOBAL_CONF_SPI_DMA_CTRL_DST_SRAM_SHIFT
 #define SPI_DMA_STATUS_ERROR_MASK		PSOC_GLOBAL_CONF_SPI_DMA_STATUS_ERROR_MASK
diff --git a/arch/gaudi3/arc/onix.c b/arch/gaudi3/arc/onix.c
index 987fd22..17657e0 100644
--- a/arch/gaudi3/arc/onix.c
+++ b/arch/gaudi3/arc/onix.c
@@ -7,6 +7,13 @@
 #include <arcboot.h>
 #include <memory.h>
 
+#define SYS_CLOCK_FREQUENCY	35000000ULL
+
+uint64_t arcboot_rom_get_timestamp_freq(void)
+{
+	return SYS_CLOCK_FREQUENCY;
+}
+
 /*
  * Because of ONIX RAM size and layout limitation,
  * both MEM_LKD_PCI and MEM_FWUPD_PCI, are overlapped.
diff --git a/arch/gaudi3/arc/regmap.h b/arch/gaudi3/arc/regmap.h
index 5ad5cfe..614fdf7 100644
--- a/arch/gaudi3/arc/regmap.h
+++ b/arch/gaudi3/arc/regmap.h
@@ -21,6 +21,12 @@
 #define mmEFUSE_READ_ADDRESS		mmD0_PSOC_EFUSE_READ_ADDRESS
 #define mmEFUSE_DATA_0			mmD0_PSOC_EFUSE_DATA_0
 
+#define mmTIMESTAMP_CNTCR		mmD0_PSOC_TIMESTAMP_CNTCR
+#define mmTIMESTAMP_CNTSR		mmD0_PSOC_TIMESTAMP_CNTSR
+#define mmTIMESTAMP_CNTCVL		mmD0_PSOC_TIMESTAMP_CNTCVL
+#define mmTIMESTAMP_CNTCVU		mmD0_PSOC_TIMESTAMP_CNTCVU
+#define mmTIMESTAMP_CNTFID0		mmD0_PSOC_TIMESTAMP_CNTFID0
+
 #define mmPRSTN				mmD0_PSOC_GLOBAL_CONF_PRSTN
 #define mmCOLD_RST_FLOPS_2		mmD0_PSOC_GLOBAL_CONF_COLD_RST_FLOPS_2
 #define mmPCIE_PHY_RST_N		mmD0_PSOC_GLOBAL_CONF_PCIE_PHY_RST_N
diff --git a/arch/gaudi3/arc/regmask.h b/arch/gaudi3/arc/regmask.h
index 0bc5e6a..ce5c800 100644
--- a/arch/gaudi3/arc/regmask.h
+++ b/arch/gaudi3/arc/regmask.h
@@ -6,6 +6,9 @@
 #ifndef _GAUDI3_REGMASK_H_
 #define _GAUDI3_REGMASK_H_
 
+#define TIMESTAMP_CNTCR_CNTCR_HDBG_M		D0_CPU_TIMESTAMP_CNTCR_CNTCR_HDBG_M
+#define TIMESTAMP_CNTCR_CNTCR_EN_M		D0_CPU_TIMESTAMP_CNTCR_CNTCR_EN_M
+
 #define SPI_DMA_CTRL_ADDR_SHIFT			D0_PARC_GLOBAL_CONF_SPI_DMA_CTRL_ADDR_SHIFT_S
 #define SPI_DMA_CTRL_DST_SRAM_SHIFT		D0_PARC_GLOBAL_CONF_SPI_DMA_CTRL_DST_SRAM_S
 #define SPI_DMA_STATUS_ERROR_MASK		D0_PARC_GLOBAL_CONF_SPI_DMA_STATUS_ERROR_M
diff --git a/arch/greco/arc_elf32/regmap.h b/arch/greco/arc_elf32/regmap.h
index e3a6653..99ee1e2 100644
--- a/arch/greco/arc_elf32/regmap.h
+++ b/arch/greco/arc_elf32/regmap.h
@@ -17,6 +17,12 @@
 #define mmQSPI_BAUDR			mmPSOC_QSPI_BAUDR
 #define mmQSPI_DR0			mmPSOC_QSPI_DR0
 
+#define mmTIMESTAMP_CNTCR		mmPSOC_TIMESTAMP_CNTCR
+#define mmTIMESTAMP_CNTSR		mmPSOC_TIMESTAMP_CNTSR
+#define mmTIMESTAMP_CNTCVL		mmPSOC_TIMESTAMP_CNTCVL
+#define mmTIMESTAMP_CNTCVU		mmPSOC_TIMESTAMP_CNTCVU
+#define mmTIMESTAMP_CNTFID0		mmPSOC_TIMESTAMP_CNTFID0
+
 #define mmEFUSE_READ_CTRL		mmPSOC_EFUSE_READ
 #define mmEFUSE_READ_ADDRESS		mmPSOC_EFUSE_READ_ADDRESS
 #define mmEFUSE_DATA_0			mmPSOC_EFUSE_DATA_0
diff --git a/arch/greco/arc_elf32/regmask.h b/arch/greco/arc_elf32/regmask.h
index 1e9c392..0ed584a 100644
--- a/arch/greco/arc_elf32/regmask.h
+++ b/arch/greco/arc_elf32/regmask.h
@@ -6,6 +6,9 @@
 #ifndef _GRECO_REGMASK_H_
 #define _GRECO_REGMASK_H_
 
+#define TIMESTAMP_CNTCR_CNTCR_HDBG_M		CPU_TIMESTAMP_CNTCR_CNTCR_HDBG_MASK
+#define TIMESTAMP_CNTCR_CNTCR_EN_M		CPU_TIMESTAMP_CNTCR_CNTCR_EN_MASK
+
 #define SPI_DMA_CTRL_ADDR_SHIFT			PSOC_GLOBAL_CONF_SPI_DMA_CTRL_ADDR_SHIFT
 #define SPI_DMA_CTRL_DST_SRAM_SHIFT		PSOC_GLOBAL_CONF_SPI_DMA_CTRL_DST_SRAM_SHIFT
 #define SPI_DMA_STATUS_ERROR_MASK		PSOC_GLOBAL_CONF_SPI_DMA_STATUS_ERROR_MASK
diff --git a/common/delay.c b/common/delay.c
new file mode 100644
index 0000000..f9d9cbc
--- /dev/null
+++ b/common/delay.c
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <arcboot.h>
+
+#define SYS_REF_CLOCK		50000000ULL
+#define USEC_PER_SEC		1000000
+#define BYPASS_CYCLES_PER_USEC	(SYS_REF_CLOCK / USEC_PER_SEC)
+
+static uint64_t cycles_per_usec = BYPASS_CYCLES_PER_USEC;
+
+__weak uint64_t arcboot_rom_get_timestamp_freq(void)
+{
+	return SYS_REF_CLOCK;
+}
+
+void arcboot_rom_init_ts(void)
+{
+	uint32_t timestamp_freq = arcboot_rom_get_timestamp_freq();
+
+	cycles_per_usec = timestamp_freq / USEC_PER_SEC;
+
+	/* timer init */
+	regmap_writel(timestamp_freq, mmTIMESTAMP_CNTFID0);
+	regmap_writel((TIMESTAMP_CNTCR_CNTCR_EN_M | TIMESTAMP_CNTCR_CNTCR_HDBG_M),
+			mmTIMESTAMP_CNTCR); /* EN | HDBG */
+}
+
+static uint64_t udelay_get_ts(void)
+{
+	uint32_t low, up1, up2;
+
+	up1 = regmap_readl(mmTIMESTAMP_CNTCVU);
+	low = regmap_readl(mmTIMESTAMP_CNTCVL);
+	up2 = regmap_readl(mmTIMESTAMP_CNTCVU);
+	/* In case of overlap read low part again */
+	if (up1 != up2)
+		low = regmap_readl(mmTIMESTAMP_CNTCVL);
+
+	return ((uint64_t)up2 << 32) + low;
+}
+
+uint64_t arcboot_rom_get_ts_us(void)
+{
+	return udelay_get_ts() / cycles_per_usec;
+}
+
+bool arcboot_rom_timeout_expired(uint64_t start_tv, uint32_t delay)
+{
+	uint64_t current, timeout_tv = start_tv + delay;
+	bool overlap = false;
+
+	if (timeout_tv < start_tv)
+		overlap = true;
+
+	current = arcboot_rom_get_ts_us();
+	if (!overlap) {
+		if (current > timeout_tv || current <= start_tv)
+			return true;
+	} else {
+		if (current > timeout_tv && current < start_tv)
+			return true;
+	}
+
+	return false;
+}
diff --git a/include/delay.h b/include/delay.h
index 0d5b67c..c9c5225 100644
--- a/include/delay.h
+++ b/include/delay.h
@@ -24,4 +24,8 @@
         : "lp_count");
 }
 
+void arcboot_rom_init_ts(void);
+bool arcboot_rom_timeout_expired(uint64_t start_tv, uint32_t delay);
+uint64_t arcboot_rom_get_ts_us(void);
+
 #endif /* _DELAY_H_ */
